### ISSUE TITLE
[public-1.14.0] Regression: AI thinking speed

### ISSUE THREAD

**Issue by jacek2023** [nmrih/source-game#1542]
Me, and many people are upset at how the new AI is unfortunately, extremely buggy, and sometimes even frustrating to play against. Objectively, the old AI was miles better than the one we have currently.

To list a few issues with the New AI:
In a group/horde, the zombies tend to not collide properly.
In areas where the zombies cannot path to the player, they tend to twitch and tweak like a crackhead. And I don't mean it in a good way. I've also noticed zombies in a group easily get stuck.
The movement is a lot more weirder, to say the least. I've heard of someone describing how a zombie did a wholeass 360 on him which is like, it shouldn't do that?
Last but not least, the AI also updates too fast.

Please, for the love of god, bug test your new AI before pushing it to the public. I don't mean this entire report as offensive, I genuinely want the best for NMRIH, which is why it upsets me so much that you push out buggy features. The easiest, and safest solution to fix these issues, is to simply revert to the old AI. You'd be making lots of players happy.

**Comment by felis-catus** [comment]
There's a lot to chew here, I'll start by saying that none of the breaking changes were intentional, it's all side effects from fixing logic which has been misbehaving for over a decade.

### What changed?
The AI itself wasn't changed, some of the previously single player exclusive logic were fixed to account for multiple players, this was done in order to fix #1475, where zombies wouldn't react in time when multiple players were active on the server.

Basically; zombies now behave in multiplayer as they did in solo, and the repercussions of it are now seen far and wide.

Technically; this was broken in the base SDK code, single player logic kept throttling zombie thinking / sensing frequency when multiple clients were present, each player took turns every 100ms to be the definitive "single player", and test if they were potentially visible; one player residing or spectating at inactive section of the map would cause all zombies to reduce their thinking speed. Having more players resulted in more iteration = slower thinking, which caused the aforementioned aggro bug.
Therefore, optimization tests were rewritten to explicitly test nearest alive player and make assumptions based on that.

Because of this fundamentally broken logic, AI relied on undefined behavior across different scenarios, unintentionally altered by player count and their activity state. For consistency's sake, it's better to have actual solutions for the problems at hand.

### To answer some points

> In a group/horde, the zombies tend to not collide properly.

Is there a way to spot or reproduce this? Collision logic or "greasing" itself wasn't altered.

> In areas where the zombies cannot path to the player, they tend to twitch and tweak like a crackhead. And I don't mean it in a good way. I've also noticed zombies in a group easily get stuck.

True, their faster thinking speed made an existing path failure loop worse, causing them to seek new targets more rapidly. Adding defined target selection delay would be a potential solution for now.

> The movement is a lot more weirder, to say the least. I've heard of someone describing how a zombie did a wholeass 360 on him which is like, it shouldn't do that?

This may also be caused by their increased thinking speed, needs more investigation if their rotation is tied to such. I think they previously "stumbled" turns due to their updates cutting short. In multiplayer this can be more apparent due to their increased targeting rate.

> Last but not least, the AI also updates too fast.

This is the root problem, active NPCs now think at a consistent rate, we'll have to explicitly delay their actions.


### Responding to QA concerns
The community team is very small, and multiplayer testing is usually limited to few cases with bots. Sometimes "trial by fire" is an necessary evil to test remaining real world scenarios, it's an overwhelming effort to test different cases, server configurations, and gameplay scenarios. The fix for #1475 is a text-book example of something that looks simple on paper, but ended up unearthing this major bug, and the fact that the game has been running undefined zombie behavior for this past 10+ years. Needless to say, the urge to fix such an outstanding bug was high, and we decided to investigate and take action, even with the limited resources.

The community maintainer team also genuinely wants the best for NMRiH, while the aggro fix may have been the catalyst, the intentions were always good; it was to fix an outstanding issue affecting the game. Software development is difficult, it's why companies got hundreds and thousands of people shipping products, it's a lot different for few spare-time developers.

In conclusion, since the problems listed can be pinpointed to their thinking speed and re-targeting rate, I'll rename this issue and track it accordingly. Thanks for contributing.

### tl;dr it'll be fixed

_(speculative fixes for idle zombies applied in src-r7768 and src-r7769)_

**Comment by felis-catus** [comment]
Twitching and tweaking fixed in src-r7782.

**Comment by felis-catus** [comment]
Rest of the issues should be fixed in commit range src-r7787-r7792.

In addition, resolved a problem where zombie animations would become jittery after all players entered spectator mode (and in some other cases); the efficiency rules were still written with the broken visibility behavior in mind, where clients took turns.

Greasing had a consistency issue due to zombies not entering their "angry" wandering routine when an enemy became unreachable (it's where the greasing task occurs), instead they reverted to "medium" wander; this regression has been in the build for quite some time (at least since 1.13.x?), and has now been fixed; they appear to switch collision groups more often.

Keeping this issue open until we can review the fixes.

**Comment by felis-catus** [comment]
Closing as resolved, all listed issues have been dealt with. If the fixes won't cut it, 1.14.1 will also include `ai_legacy` cvar to run bugged behavior; obviously unsupported and not recommended, but it's there for legacy version 'emulation' when required.
